Isidora Popovic \
2024/3285\
Zadatak 2 Vremenska prognoza

Visok pritisak više sunca\
Nizak pritisak  ili sunce ili kiša\
Za rešavanje ovog problema koristićemo Markovljev lanac, gde temperatura sutra zavisi od svih temperatura do sada

In [17]:
import numpy as np
import matplotlib.pyplot as plt
import random
import pandas as pd 
import os 

In [18]:
P = np.array([
    [0.9,  0.0,  0.1],
    [0.05, 0.5,  0.45],
    [0.05, 0.7,  0.25]
])

In [19]:
pi0 = np.array([0.10, 0.70, 0.20])

In [20]:
states = ["V-sunčano", "N-sunčano", "N-kišovito"]
df_P = pd.DataFrame(P, index=states, columns=states)
df_pi0 = pd.DataFrame(pi0.reshape(1,-1), columns=states, index=["1. jun"])

In [21]:
print("Tranziciona matrica P:")
display(df_P)

Tranziciona matrica P:


,V-sunčano,N-sunčano,N-kišovito
V-sunčano,0.90,0.0,0.10
N-sunčano,0.05,0.5,0.45
N-kišovito,0.05,0.7,0.25


In [22]:
print("Početna raspodela pi0 na 1. jun:")
display(df_pi0)

Početna raspodela pi0 na 1. jun:


,V-sunčano,N-sunčano,N-kišovito
1. jun,0.1,0.7,0.2


In [23]:
def verovatnoca(pi_initial, P, n):
    Pn = np.linalg.matrix_power(P, n)  # P^n
    return pi_initial.dot(Pn)

In [24]:
p6jun=verovatnoca(pi0, P, 5)
p1jul=verovatnoca(pi0, P, 30)

In [25]:
print(f"Raspodela na 6. jun ({p6jun} dana kasnije):")
print(f"  V-sunčano:   {p6jun[0]:.2f}")
print(f"  N-sunčano:   {p6jun[1]:.2f}")
print(f"  N-kišovito:  {p6jun[2]:.2f}")

print(f"\nRaspodela na 1. jul ({p1jul} dana kasnije):")
print(f"  V-sunčano:   {p1jul[0]:.3f}")
print(f"  N-sunčano:   {p1jul[1]:.3f}")
print(f"  N-kišovito:  {p1jul[2]:.3f}")

Raspodela na 6. jun ([0.22980209 0.45785994 0.31233797] dana kasnije):
  V-sunčano:   0.23
  N-sunčano:   0.46
  N-kišovito:  0.31

Raspodela na 1. jul ([0.33155282 0.3900759  0.27837128] dana kasnije):
  V-sunčano:   0.332
  N-sunčano:   0.390
  N-kišovito:  0.278


In [26]:
pi0 = np.array([0.1, 0.7, 0.2], dtype=float)

def raspodela_za(pi0, P, n):
    return pi0 @ np.linalg.matrix_power(P, n)

pi_5  = raspodela_za(pi0, P, 5)
pi_30 = raspodela_za(pi0, P, 30)

print("Za 5 dana :", np.round(pi_5, 4))
print("Za 30 dana:", np.round(pi_30, 4))

Za 5 dana : [0.2298 0.4579 0.3123]
Za 30 dana: [0.3316 0.3901 0.2784]


In [27]:
for dan in range(1, 7):
    n = dan - 1
    pin = raspodela_za(pi0, P, n)
    df = pd.DataFrame([np.round(pin, 4)], index=[f"{dan}. jun"], columns=states)
    print(f"\n=== {dan}. jun — raspodela stanja ===")
    print(df.to_string())


=== 1. jun — raspodela stanja ===
        V-sunčano  N-sunčano  N-kišovito
1. jun        0.1        0.7         0.2

=== 2. jun — raspodela stanja ===
        V-sunčano  N-sunčano  N-kišovito
2. jun      0.135       0.49       0.375

=== 3. jun — raspodela stanja ===
        V-sunčano  N-sunčano  N-kišovito
3. jun     0.1648     0.5075      0.3278

=== 4. jun — raspodela stanja ===
        V-sunčano  N-sunčano  N-kišovito
4. jun       0.19     0.4832      0.3268

=== 5. jun — raspodela stanja ===
        V-sunčano  N-sunčano  N-kišovito
5. jun     0.2115     0.4703      0.3181

=== 6. jun — raspodela stanja ===
        V-sunčano  N-sunčano  N-kišovito
6. jun     0.2298     0.4579      0.3123


Kako vreme prolazi, vidimo da se sve više približavamo uniformnoj raspodeli — verovatnoća za svako vreme iznosi oko 33%, jer nam ulazni podaci slabe i ne donose nikakvu novu informaciju. Ako je danas bilo sunčano, očekujemo da će i sutra biti sunčano. Promena verovatnoće tranzicije prati i verovatnoću pojave tog vremena. Na primer, ako smanjimo verovatnoću da visok pritisak rezultuje sunčanim vremenom, ne možemo biti sigurni da će sutra biti sunčano

Gates, P., and H. Tong. "On Markov chain modeling to some weather data." Journal of Applied Meteorology and Climatology 15.11 (1976): 1145-1151.\
https://medium.com/data-science/predicting-the-weather-with-markov-chains-a34735f0c4df